### PR TITLE
MULE-15009: JavaModule: Support FQN for parameter types to identify methods

### PR DESCRIPTION
**Pull request by ndinu**

### CODE REVIEW

File: src/main/java/org/mule/extensions/java/internal/parameters/ConstructorIdentifier.java [hunk 1]
```diff
@@ -70,7 +70,9 @@ public String getElementId() {
 
   @Override
   public boolean matches(Executable element) {
-    return element instanceof Constructor && new ConstructorIdentifier((Constructor) element).equals(this);
+    return element instanceof Constructor
+        && getElementName().equals(element.getDeclaringClass().getSimpleName())
+        && matchesArguments(element.getParameterTypes());
```
Comment (alegmarra) [inline src/main/java/org/mule/extensions/java/internal/parameters/ConstructorIdentifier.java]: this does not look good, the equality of the identifier should be enough, why falling back to each part separately?

File: src/main/java/org/mule/extensions/java/internal/parameters/ExecutableIdentifier.java [hunk 2]
```diff
@@ -58,6 +62,27 @@ public String getElementName() {
     return match.matches() ? match.group(1) : "";
   }
 
+  /**
+   * Provides the type name of the {@link Executable} arguments being represented by {@code this} {@link ExecutableIdentifierFactory}.
+   * <p>
+   * For example, for method {@code public void echo(String message, int times)}, the result of invoking this method will be
+   * {@code ["String", "int"]}.
+   *
+   * @return the name of the {@link Executable} element being represented by {@code this} {@link ExecutableIdentifierFactory}.
```
Comment (alegmarra) [inline src/main/java/org/mule/extensions/java/internal/parameters/ExecutableIdentifier.java]: what's the difference between Optional.empty and empty Array?

Comment (alegmarra) [reply src/main/java/org/mule/extensions/java/internal/parameters/ExecutableIdentifier.java]: also, this is not returning the name of the element, but the name of its arguments

File: src/main/java/org/mule/extensions/java/internal/parameters/MethodIdentifier.java [hunk 3]
```diff
@@ -76,7 +76,8 @@ public String getElementId() {
   public boolean matches(Executable element) {
     return element instanceof Method
         && !isStatic(element.getModifiers())
-        && this.equals(new MethodIdentifier((Method) element));
+        && getElementName().equals(element.getName())
```
Comment (alegmarra) [inline src/main/java/org/mule/extensions/java/internal/parameters/MethodIdentifier.java]: same question

File: src/test/java/org/mule/extensions/internal/metadata/KeysMetadataTestCase.java [hunk 4]
```diff
@@ -58,4 +82,9 @@ public void constructorKeys() throws Exception {
         .collect(Collectors.toList());
   }
 
+  private List<String> getDisplayNames(Set<MetadataKey> keys) {
+    return keys.stream().flatMap(k -> k.getChilds().stream())
+        .map(MetadataKey::getDisplayName)
+        .collect(Collectors.toList());
```
Comment (estebanwasinger) [inline src/test/java/org/mule/extensions/internal/metadata/KeysMetadataTestCase.java]: static import `toList()`

File: src/main/java/org/mule/extensions/java/internal/metadata/ExecutableElementTypeResolver.java [hunk 5]
```diff
@@ -135,18 +143,51 @@ protected MetadataKey buildMethodKeys(String clazz) throws MetadataResolvingExce
...
+  }
+
+  private boolean parameterTypesWithinExecutableClash(Executable executable) {
+    Class[] parameterTypes = executable.getParameterTypes();
+    Map<String, Set<String>> parameterTypesMap = new HashMap<>();
+    for (Class type : parameterTypes) {
+      parameterTypesMap.computeIfAbsent(type.getSimpleName(), key -> new HashSet<>()).add(type.getCanonicalName());
+    }
+
+    return parameterTypesMap.values().stream().anyMatch(set -> set.size() > 1);
```
Comment (alegmarra) [inline src/main/java/org/mule/extensions/java/internal/metadata/ExecutableElementTypeResolver.java]:
```
Map<String, String> paramTypes = new HashMap<>();
for (Class type: executable.getParameterTypes()){
  String canonicalName = type.getCanonicalName();
  String boundName = paramTypes.computeIfAbsent(type.getSimpleName(), k -> canonicalName);
  if (!canonicalName.equals(boundName)){
    return true;
  } 
}
return false;
```

**Pull request merged**